tests/executables_src/testLMapMathPluginPushPars.cc

```diff
@@ -0,0 +1,125 @@
+    // - therefore, in test here, we must issue write for all push-parameters
+    // - then, version number check in MathPlugin will notice valid data has been provided after open
+    // - for poll-type accessors, version number check also succeeds since they get recent version numbers anyway
+    // Test needs to call activateAsyncRead but it must be irrelevant whether before or after writes
+
+    ChimeraTK::Device targetDevice;
+    targetDevice.open("HOLD");
+
+    auto accTarget = targetDevice.getScalarRegisterAccessor<uint32_t>("HOLD0/WORD_G");
+    auto pollPar = targetDevice.getScalarRegisterAccessor<uint32_t>("HOLD0/POLLPAR");
+    pollPar = 1;
+    pollPar.write();
+
+    ChimeraTK::Device device;
+    device.open("EOD");
+    device.activateAsyncRead();
+    auto pushPar = device.getScalarRegisterAccessor<uint32_t>("DET/EXPOSURE");
+
+    pushPar = 2;
+    pushPar.write();
+
+    auto accMathWrite = device.getScalarRegisterAccessor<double>("DET/GAIN");
+    // we don't have main value (x in formula) yet, since it wasn't yet written.
+    // therefore, we expect to have no value yet for formula output (0 is default from dummy construction)
+    accTarget.readLatest();
+    BOOST_CHECK_EQUAL(int(accTarget), 0);
+
+    // write to main value and check result
+    accMathWrite = 3;
+    accMathWrite.write();
+    accTarget.readLatest();
+    BOOST_CHECK_EQUAL(int(accTarget), 100 * (int)pollPar + 10 * (int)pushPar + (int)accMathWrite);
+
+    // write to push-parameter and check result
+    // note, it's a new feature that result is completely written when write() returns.
+    pushPar = 4;
+    pushPar.write();
+    accTarget.readLatest();
+    BOOST_CHECK_EQUAL(int(accTarget), 100 * (int)pollPar + 10 * (int)pushPar + (int)accMathWrite);
+
+    // check that MathPlugin waits on all initial values.
+    device.close();
+    targetDevice.open(); // open again since low-level device was closed by LNM
+    accTarget = 0;       // reset result in dummy
+    accTarget.write();
+    pollPar.write(); // restore value in case lost during reopen (actually not required with current dummy impl)
+    device.open("EOD");
+    device.activateAsyncRead();
+
+    accMathWrite = 5;
+    accMathWrite.write();
+    // check that MathPlugin did not yet write to the device - it must wait on push-parameter value
+    accTarget.readLatest();
+    BOOST_CHECK_EQUAL(int(accTarget), 0);
+
+    pushPar.write();
+    accTarget.readLatest();
+    BOOST_CHECK_EQUAL(int(accTarget), 100 * (int)pollPar + 10 * (int)pushPar + (int)accMathWrite);
+
+    // user expectation will be that write-behavior does not depend on when we call activateAsyncRead,
+    // so test that update is retrieved even if it's called last
+    device.close();
+    device.open();
+    accTarget = 0;
+    accTarget.write();
+    pollPar.write();
+    accMathWrite = 7;
+    accMathWrite.write();
+    pushPar = 6;
+    pushPar.write();
+    device.activateAsyncRead();
+    accTarget.readLatest();
```

Comment:
What actually is missing is the test that write is only happening once. When reading though the code I thought it would be executed multiple times. It is not, but there is no check here. You could extend the dummy with a write counter on the target register.

Reply:
Yes, adding this check actually showed a bug. 
But it was not complicated to fix. Only needed to replace list(Math Plugins) -> set(Math Plugins)